core/src/main/scala/kafka/log/Log.scala

```diff
@@ -243,6 +269,26 @@ class Log(val dir: File,
       recoverLog()
       // reset the index size of the currently active log segment to allow more entries
       activeSegment.index.resize(config.maxIndexSize)
+      activeSegment.timeIndex.resize(config.maxIndexSize)
+    }
+
+    /**
+     * We append a time index entry for non-active segment with an empty time index. This could happen during upgrade.
+     * We need to do this because searching by timestamp needs a time index entry.
+     */
+    val logSegmentsSeq = logSegments.toSeq
+    for (i <- 0 until logSegmentsSeq.size) {
+      // Insert the last entry to the time index for inactive segments.
+      if (logSegmentsSeq(i) != activeSegment) {
+        // It is possible that the time index was created for the recovered or swapped log segment file and is
+        // resized to 0. We need to resize it to hold at least one entry.
+        if (logSegmentsSeq(i).timeIndex.maxEntries == 0)
```

Comment:
This is to make sure the time index file is not resized to zero so it can hold at least one time index entry. But putting the check here is a little confusing, I will move it into ensureNonEmptyTimeIndex().